internal/data/disbursement_instructions.go

```diff
@@ -67,166 +85,273 @@ var (
+		}
+		instruction := instructionsByContactMap[contact]
+		if instruction == nil {
+			return fmt.Errorf("instruction not found for receiver with ID %s", receiver.ID)
 		}
+		verification, exists := verificationByReceiverIDMap[receiver.ID]
 
-		// Step 5: Create payments for all receivers
-		payments := make([]PaymentInsert, 0, len(instructions))
-		for _, instruction := range instructions {
-			receiver := receiverMap[instruction.Phone]
-			payment := PaymentInsert{
+		if !exists {
+			verificationInsert := ReceiverVerificationInsert{
 				ReceiverID:        receiver.ID,
-				DisbursementID:    disbursement.ID,
-				Amount:            instruction.Amount,
-				AssetID:           disbursement.Asset.ID,
-				ReceiverWalletID:  receiverIDToReceiverWalletIDMap[receiver.ID],
-				ExternalPaymentID: instruction.ExternalPaymentId,
+				VerificationValue: instruction.VerificationValue,
+				VerificationField: disbursement.VerificationField,
+			}
+			_, insertErr := di.receiverVerificationModel.Insert(ctx, dbTx, verificationInsert)
+			if insertErr != nil {
+				return fmt.Errorf("error inserting receiver verification: %w", insertErr)
+			}
+		} else if !CompareVerificationValue(verification.HashedValue, instruction.VerificationValue) {
+			if verification.ConfirmedAt != nil {
+				return fmt.Errorf("%w: receiver verification for %s doesn't match. Check instruction with ID %s", ErrReceiverVerificationMismatch, contact, instruction.ID)
+			}
+			updateErr := di.receiverVerificationModel.UpdateVerificationValue(ctx, dbTx, verification.ReceiverID, verification.VerificationField, instruction.VerificationValue)
+			if updateErr != nil {
+				return fmt.Errorf("error updating receiver verification for disbursement id %s: %w", disbursement.ID, updateErr)
 			}
-			payments = append(payments, payment)
-		}
-		if err = di.paymentModel.InsertAll(ctx, dbTx, payments); err != nil {
-			return fmt.Errorf("error inserting payments: %w", err)
 		}
+	}
 
-		// Step 6: Persist Payment file to Disbursement
-		if err = di.disbursementModel.Update(ctx, update); err != nil {
-			return fmt.Errorf("error persisting payment file: %w", err)
+	return nil
+}
+
+func (di DisbursementInstructionModel) processReceiverWallets(ctx context.Context, dbTx db.DBTransaction, receiverMap map[string]*Receiver, disbursement *Disbursement) (map[string]string, error) {
+	receiverIDs := make([]string, 0, len(receiverMap))
+	for id := range receiverMap {
+		receiverIDs = append(receiverIDs, id)
+	}
+
+	receiverWallets, err := di.receiverWalletModel.GetByReceiverIDsAndWalletID(ctx, dbTx, receiverIDs, disbursement.Wallet.ID)
+	if err != nil {
+		return nil, fmt.Errorf("fetching receiver wallets: %w", err)
+	}
+	receiverIDToReceiverWalletIDMap := make(map[string]string)
+	for _, receiverWallet := range receiverWallets {
+		receiverIDToReceiverWalletIDMap[receiverWallet.Receiver.ID] = receiverWallet.ID
+	}
+
+	for receiverId := range receiverMap {
```

Comment:
💡 According to the Go naming conventions, acronyms should be either all-uppercased or all-lowercased. Plz, refer to the initialisms section from Go's official [CodeReviewComments wiki](https://go.dev/wiki/CodeReviewComments).

```suggestion
	for receiverID := range receiverMap {
```

Reply:
I just refactored this code by splitting it into methods. I tried to fix some issues during the refactor but didn't catch everything.

Reply:
addressed in e3c251055d56b06803dac1f04d4c516f20d37337